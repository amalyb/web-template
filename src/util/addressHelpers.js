/**
 * Address normalization and mapping utilities
 */

// Normalize address data by trimming spaces and converting to standard formats
export const normalizeAddress = (address) => {
  if (!address) return null;

  const normalized = {
    fullName: address.fullName?.trim().replace(/\s+/g, ' ') || '',
    country: address.country?.toUpperCase() || 'US',
    street1: address.street1?.trim().replace(/\s+/g, ' ') || '',
    street2: address.street2?.trim().replace(/\s+/g, ' ') || undefined, // Always send undefined when missing
    city: address.city?.trim().replace(/\s+/g, ' ') || '',
    state: address.state?.trim().toUpperCase() || '',
    zip: address.zip?.trim().replace(/\s+/g, '') || '',
    email: address.email?.trim().toLowerCase() || '',
    phone: normalizePhone(address.phone) || '',
  };

  // Clean up empty strings to undefined for optional fields
  if (!normalized.street2) normalized.street2 = undefined;
  if (!normalized.phone) normalized.phone = undefined;

  return normalized;
};

// Normalize phone number to E.164 format
export const normalizePhone = (phone) => {
  if (!phone) return '';
  
  // Remove extensions (x123, ext 123, etc.) and reject them
  let cleaned = phone.replace(/\s*(x|ext|extension)\s*\d+.*$/i, '');
  
  // Remove all non-digit characters except +
  cleaned = cleaned.replace(/[^\d+]/g, '');
  
  // If it doesn't start with +, assume US number and add +1
  if (!cleaned.startsWith('+')) {
    // Remove leading 1 if present
    if (cleaned.startsWith('1') && cleaned.length === 11) {
      cleaned = cleaned.substring(1);
    }
    cleaned = '+1' + cleaned;
  }
  
  // Validate E.164 format (1-15 digits after +)
  const e164Regex = /^\+[1-9]\d{1,14}$/;
  if (!e164Regex.test(cleaned)) {
    return phone; // Return original if normalization fails
  }
  
  return cleaned;
};

// Map normalized address to Stripe billing format
export const mapToStripeBilling = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address: {
      line1: normalized.street1,
      line2: normalized.street2 || undefined, // Always send undefined when missing
      city: normalized.city,
      state: normalized.state,
      postal_code: normalized.zip,
      country: normalized.country,
    },
  };
};

// Map normalized address to Shippo/UPS/FedEx format
export const mapToShippo = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    street1: normalized.street1,
    street2: normalized.street2 || '',
    city: normalized.city,
    state: normalized.state,
    zip: normalized.zip,
    country: normalized.country,
  };
};

// Check if address is a PO Box (for courier routing)
export const isPOBox = (address) => {
  if (!address?.street1) return false;
  return /^(P(OST)?\.?\s*O(FFICE)?\.?\s*BOX)\b/i.test(address.street1);
};

// Map normalized address to UPS format
export const mapToUPS = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    name: normalized.fullName,
    email: normalized.email,
    phone: normalized.phone,
    address_line1: normalized.street1,
    address_line2: normalized.street2 || '',
    city: normalized.city,
    state_province_code: normalized.state,
    postal_code: normalized.zip,
    country_code: normalized.country,
  };
};

// Map normalized address to FedEx format
export const mapToFedEx = (address) => {
  const normalized = normalizeAddress(address);
  if (!normalized) return null;

  return {
    contact: {
      person_name: normalized.fullName,
      email_address: normalized.email,
      phone_number: normalized.phone,
    },
    address: {
      street_lines: [normalized.street1, normalized.street2].filter(Boolean),
      city: normalized.city,
      state_or_province_code: normalized.state,
      postal_code: normalized.zip,
      country_code: normalized.country,
    },
  };
};

// Validate address completeness
export const validateAddress = (address, requiredFields = {}) => {
  const defaults = {
    fullName: true,
    country: true,
    street1: true,
    city: true,
    state: true,
    zip: true,
    email: true,
    phone: false,
  };
  
  const requirements = { ...defaults, ...requiredFields };
  const errors = {};
  
  if (requirements.fullName && !address.fullName?.trim()) {
    errors.fullName = 'Full name is required';
  }
  
  if (requirements.country && !address.country?.trim()) {
    errors.country = 'Country is required';
  }
  
  if (requirements.street1 && !address.street1?.trim()) {
    errors.street1 = 'Street address is required';
  }
  
  if (requirements.city && !address.city?.trim()) {
    errors.city = 'City is required';
  }
  
  if (requirements.state && !address.state?.trim()) {
    errors.state = 'State/Region is required';
  }
  
  if (requirements.zip && !address.zip?.trim()) {
    errors.zip = 'ZIP/Postal code is required';
  }
  
  if (requirements.email && !address.email?.trim()) {
    errors.email = 'Email is required';
  } else if (address.email && !/^[^\s@]+@[^\s@]+\.[^\s@]+$/.test(address.email)) {
    errors.email = 'Please enter a valid email address';
  }
  
  if (requirements.phone && !address.phone?.trim()) {
    errors.phone = 'Phone number is required';
  }
  
  return {
    isValid: Object.keys(errors).length === 0,
    errors,
  };
};

// Copy billing address to shipping address
export const copyBillingToShipping = (billingAddress) => {
  const normalized = normalizeAddress(billingAddress);
  if (!normalized) return null;

  return {
    shipping: {
      fullName: normalized.fullName,
      country: normalized.country,
      street1: normalized.street1,
      street2: normalized.street2,
      city: normalized.city,
      state: normalized.state,
      zip: normalized.zip,
      email: normalized.email,
      phone: normalized.phone,
    },
  };
};
