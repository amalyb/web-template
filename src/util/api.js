// These helpers are calling this template's own server-side routes
// so, they are not directly calling Marketplace API or Integration API.
// You can find these api endpoints from 'server/api/...' directory

import axios from 'axios';
import appSettings from '../config/settings';
import { types as sdkTypes, transit } from './sdkLoader';
import Decimal from 'decimal.js';

export const apiBaseUrl = marketplaceRootURL => {
  const port = process.env.REACT_APP_DEV_API_SERVER_PORT;
  const useDevApiServer = process.env.NODE_ENV === 'development' && !!port;

  // In development, the dev API server is running in a different port
  if (useDevApiServer) {
    return `http://localhost:${port}`;
  }

  // Otherwise, use the given marketplaceRootURL parameter or the same domain and port as the frontend
  return marketplaceRootURL ? marketplaceRootURL.replace(/\/$/, '') : `${window.location.origin}`;
};

// Axios client for API calls with Basic authentication
const apiBase = process.env.REACT_APP_API_BASE_URL || '/api';
export const apiClient = axios.create({ baseURL: apiBase });

apiClient.interceptors.request.use((config) => {
  const u = process.env.REACT_APP_BASIC_AUTH_USERNAME;
  const p = process.env.REACT_APP_BASIC_AUTH_PASSWORD;
  if (u && p) {
    const token = btoa(`${u}:${p}`);
    config.headers = { ...config.headers, Authorization: `Basic ${token}` };
  }
  return config;
});

// Application type handlers for JS SDK.
//
// NOTE: keep in sync with `typeHandlers` in `server/api-util/sdk.js`
export const typeHandlers = [
  // Use Decimal type instead of SDK's BigDecimal.
  {
    type: sdkTypes.BigDecimal,
    customType: Decimal,
    writer: v => new sdkTypes.BigDecimal(v.toString()),
    reader: v => new Decimal(v.value),
  },
];

const serialize = data => {
  return transit.write(data, { typeHandlers, verbose: appSettings.sdk.transitVerbose });
};

const deserialize = str => {
  return transit.read(str, { typeHandlers });
};

const methods = {
  POST: 'POST',
  GET: 'GET',
  PUT: 'PUT',
  PATCH: 'PATCH',
  DELETE: 'DELETE',
};

// If server/api returns data from SDK, you should set Content-Type to 'application/transit+json'
const request = (path, options = {}) => {
  const url = `${apiBaseUrl()}${path}`;
  const { credentials, headers, body, ...rest } = options;

  // If headers are not set, we assume that the body should be serialized as transit format.
  const shouldSerializeBody =
    (!headers || headers['Content-Type'] === 'application/transit+json') && body;
  const bodyMaybe = shouldSerializeBody ? { body: serialize(body) } : {};

  const fetchOptions = {
    credentials: credentials || 'include',
    // Since server/api mostly talks to Marketplace API using SDK,
    // we default to 'application/transit+json' as content type (as SDK uses transit).
    headers: headers || { 'Content-Type': 'application/transit+json' },
    ...bodyMaybe,
    ...rest,
  };

  return window.fetch(url, fetchOptions).then(res => {
    const contentTypeHeader = res.headers.get('Content-Type');
    const contentType = contentTypeHeader ? contentTypeHeader.split(';')[0] : null;

    if (res.status >= 400) {
      return res.json().then(data => {
        let e = new Error();
        e = Object.assign(e, data);

        throw e;
      });
    }
    if (contentType === 'application/transit+json') {
      return res.text().then(deserialize);
    } else if (contentType === 'application/json') {
      return res.json();
    }
    return res.text();
  });
};

// Keep the previous parameter order for the post method.
// For now, only POST has own specific function, but you can create more or use request directly.
export const post = (path, body, options = {}) => {
  const requestOptions = {
    ...options,
    method: methods.POST,
    body,
  };

  return request(path, requestOptions);
};

// Fetch transaction line items from the local API endpoint.
//
// See `server/api/transaction-line-items.js` to see what data should
// be sent in the body.
export const transactionLineItems = (payload) =>
  apiClient.post('/transaction-line-items', payload);

// Initiate a privileged transaction.
//
// With privileged transitions, the transactions need to be created
// from the backend. This endpoint enables sending the order data to
// the local backend, and passing that to the Marketplace API.
//
// See `server/api/initiate-privileged.js` to see what data should be
// sent in the body.
export const initiatePrivileged = body => {
  return post('/api/initiate-privileged', body);
};

// Transition a transaction with a privileged transition.
//
// This is similar to the `initiatePrivileged` above. It will use the
// backend for the transition. The backend endpoint will add the
// payment line items to the transition params.
//
// See `server/api/transition-privileged.js` to see what data should
// be sent in the body.
export const transitionPrivileged = body => {
  return post('/api/transition-privileged', body);
};

// Create user with identity provider (e.g. Facebook or Google)
//
// If loginWithIdp api call fails and user can't authenticate to Marketplace API with idp
// we will show option to create a new user with idp.
// For that user needs to confirm data fetched from the idp.
// After the confirmation, this endpoint is called to create a new user with confirmed data.
//
// See `server/api/auth/createUserWithIdp.js` to see what data should
// be sent in the body.
export const createUserWithIdp = body => {
  return post('/api/auth/create-user-with-idp', body);
};

// Ensure phone number is saved to protectedData
// This can be called after sign-up or profile updates to guarantee the phone number is stored correctly
export const ensurePhoneNumber = phoneNumber => {
  if (!phoneNumber) {
    console.warn('‚ö†Ô∏è [api] No phone number provided to ensurePhoneNumber');
    return Promise.resolve();
  }
  
  console.log('üì± [api] Ensuring phone number is saved to protectedData:', phoneNumber);
  return post('/api/ensure-phone-number', { phoneNumber })
    .then(() => {
      console.log('‚úÖ [api] Phone number ensured in protectedData');
    })
    .catch(error => {
      console.warn('‚ö†Ô∏è [api] Failed to ensure phone number:', error.message);
      throw error;
    });
};
