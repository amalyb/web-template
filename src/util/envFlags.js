/**
 * Centralized environment flags for client-side code
 * Safely handles missing process object in browser environments
 */

export const IS_PROD = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'production');
export const IS_DEV  = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'development');
export const IS_TEST = (typeof process !== 'undefined' && process.env && process.env.NODE_ENV === 'test');
export const __DEV__ = !IS_PROD;
