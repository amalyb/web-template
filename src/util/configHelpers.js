import { getSupportedProcessesInfo } from '../util/transaction';

// Generic helpers for validating config values

// Validate enum type strings (if default value is given, value is considered optional)
const validEnumString = (key, value, options, defaultOption) => {
  const isUndefined = typeof value === 'undefined';
  const isValidValue = options.includes(value);
  const isValid = (typeof defaultOption !== 'undefined' && isUndefined) || isValidValue;
  const valueMaybe = isValid ? { [key]: value || defaultOption } : {};
  return [isValid, valueMaybe];
};

// Validate boolean values (if default value is given, value is considered optional)
const validBoolean = (key, value, defaultValue) => {
  const isUndefined = typeof value === 'undefined';
  const isBoolean = typeof value == 'boolean';
  const isValid = (typeof defaultValue !== 'undefined' && isUndefined) || isBoolean;

  const validValue = isBoolean ? { [key]: value } : isValid ? { [key]: defaultValue } : {};
  return [isValid, validValue];
};

const validLabel = label => {
  const isValid = typeof label === 'string';
  const labelMaybe = isValid ? { label } : {};
  return [isValid, labelMaybe];
};

////////////////////
// Merge branding //
////////////////////

// Generate darker and lighter versions of marketplace color,
// if those values are not set by default.
// Adjusted from https://gist.github.com/xenozauros/f6e185c8de2a04cdfecf
const hexToCssHsl = (hexColor, lightnessDiff) => {
  let hex = hexColor.replace(/^#/, '');
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  const r = parseInt(hex.substr(0, 2), 16) / 255;
  const g = parseInt(hex.substr(2, 2), 16) / 255;
  const b = parseInt(hex.substr(4, 2), 16) / 255;

  // r /= 255, g /= 255, b /= 255;
  const max = Math.max(r, g, b);
  const min = Math.min(r, g, b);

  let h;
  let s;
  let l = (max + min) / 2;

  if (max == min) {
    // achromatic
    h = 0;
    s = 0;
  } else {
    const d = max - min;
    s = l > 0.5 ? d / (2 - max - min) : d / (max + min);
    switch (max) {
      case r:
        h = (g - b) / d + (g < b ? 6 : 0);
        break;
      case g:
        h = (b - r) / d + 2;
        break;
      case b:
        h = (r - g) / d + 4;
        break;
    }
    h /= 6;
  }

  h = Math.round(h * 360);
  s = Math.round(s * 100);
  l = Math.round(l * 100);

  return `hsl(${h}, ${s}%, ${l + lightnessDiff}%)`;
};

const mergeBranding = (brandingConfig, defaultBranding) => {
  const marketplaceColor = brandingConfig?.marketplaceColor || defaultBranding.marketplaceColor;
  const marketplaceColorDark =
    brandingConfig?.marketplaceColorDark ||
    defaultBranding.marketplaceColorDark ||
    (marketplaceColor ? hexToCssHsl(marketplaceColor, -10) : null);
  const marketplaceColorLight =
    brandingConfig?.marketplaceColorLight ||
    defaultBranding.marketplaceColorLight ||
    (marketplaceColor ? hexToCssHsl(marketplaceColor, 10) : null);

  return {
    marketplaceColor,
    marketplaceColorDark,
    marketplaceColorLight,
    logoImageDesktopURL: brandingConfig?.logoImageDesktopURL || defaultBranding.logoImageDesktopURL,
    logoImageMobileURL: brandingConfig?.logoImageMobileURL || defaultBranding.logoImageMobileURL,
    brandImageURL: brandingConfig?.brandImageURL || defaultBranding.brandImageURL,
    facebookImageURL: brandingConfig?.facebookImageURL || defaultBranding.facebookImageURL,
    twitterImageURL: brandingConfig?.twitterImageURL || defaultBranding.twitterImageURL,
  };
};

///////////////////
// Merge layouts //
///////////////////

const mergeLayouts = (layoutConfig, defaultLayout) => {
  const searchPageVariant = layoutConfig?.searchPageVariant || defaultLayout.searchPageVariant;
  const listingPageVariant = layoutConfig?.listingPageVariant || defaultLayout.listingPageVariant;
  const aspectWidth =
    layoutConfig?.listingImage?.aspectWidth || defaultLayout?.listingImage?.aspectWidth;
  const aspectHeight =
    layoutConfig?.listingImage?.aspectHeight || defaultLayout?.listingImage?.aspectHeight;

  const isValidSearchPageConfig = ['map', 'list'].includes(searchPageVariant);
  const isValidListingPageConfig = ['hero-image', 'full-image'].includes(listingPageVariant);

  if (!isValidSearchPageConfig) {
    console.warn('Unsupported layout option for search page detected', searchPageVariant);
  } else if (!isValidListingPageConfig) {
    console.warn('Unsupported layout option for listing page detected', listingPageVariant);
  }

  return {
    searchPageVariant: isValidSearchPageConfig ? searchPageVariant : 'list',
    listingPageVariant: isValidListingPageConfig ? listingPageVariant : 'full-image',
    listingImage: {
      aspectWidth: aspectWidth || 1,
      aspectHeight: aspectHeight || 1,
      variantPrefix: defaultLayout.listingImage.variantPrefix,
    },
  };
};

/////////////////////////////
// Validate listing config //
/////////////////////////////

const validKey = (key, allKeys) => {
  const isUniqueKey = allKeys.indexOf(key) === allKeys.lastIndexOf(key);
  return [isUniqueKey, { key }];
};

const validProcessAliases = (includeForProcessAliases, processAliasesInUse) => {
  const isUndefined = typeof includeForProcessAliases === 'undefined';
  const isArray = Array.isArray(includeForProcessAliases);
  const validatedProcessAliases = isArray
    ? includeForProcessAliases.filter(pa => processAliasesInUse.includes(pa))
    : [];

  const hasValidProcessAliases = validatedProcessAliases.length >= 0;
  const isValid = hasValidProcessAliases || isUndefined;
  const validValue = hasValidProcessAliases ? validatedProcessAliases : processAliasesInUse;
  return [isValid, { includeForProcessAliases: validValue }];
};

const isStringType = str => typeof str === 'string';
const pickOptionShapes = o => isStringType(o.option) && isStringType(o.label);

const validSchemaOptions = (schemaOptions, schemaType) => {
  const isUndefined = typeof schemaOptions === 'undefined';
  const isArray = Array.isArray(schemaOptions);
  const arrayContainsOptionShapes = isArray
    ? schemaOptions.filter(pickOptionShapes).length === schemaOptions.length
    : false;
  const shouldHaveSchemaOptions = ['enum', 'multi-enum'].includes(schemaType) && !isUndefined;

  const isValid = isUndefined || shouldHaveSchemaOptions || arrayContainsOptionShapes;
  const schemaOptionsMaybe = isArray ? { schemaOptions } : {};
  return [isValid, schemaOptionsMaybe];
};

// listingExtendedDataConfig.searchPageConfig
const filterTypes = ['SelectSingleFilter', 'SelectMultipleFilter'];
const validFilterType = (filterType, schemaType) => {
  const isEnumSchemaType = ['enum', 'multi-enum'].includes(schemaType);
  const shouldHaveFilterType = isEnumSchemaType && filterTypes.includes(filterType);
  const isValid = !isEnumSchemaType || shouldHaveFilterType;
  const filterTypeMaybe = shouldHaveFilterType ? { filterType } : {};
  return [isValid, filterTypeMaybe];
};

const searchModes = ['has_all', 'has_any'];
const validSearchMode = (searchMode, schemaType) => {
  const isMultiEnumSchemaType = schemaType === 'multi-enum';
  const isUndefined = typeof searchMode === 'undefined';
  const hasValidMode = searchModes.includes(searchMode);
  const isSearchModeValid = isMultiEnumSchemaType && (isUndefined || hasValidMode);
  const isValid = !isMultiEnumSchemaType || isSearchModeValid;
  const searchModeMaybe = isSearchModeValid ? { searchMode: searchMode || 'has_all' } : {};
  return [isValid, searchModeMaybe];
};

const validSearchPageConfig = (config, schemaType) => {
  const isUndefined = typeof config === 'undefined';
  if (isUndefined) {
    return [true, {}];
  }
  // Validate: label, filterType, searchMode, group
  const [isValidLabel, label] = validLabel(config.label);
  const [isValidFilterType, filterType] = validFilterType(config.filterType, schemaType);
  const [isValidSearchMode, searchMode] = validSearchMode(config.searchMode, schemaType);
  const groupOptions = ['primary', 'secondary'];
  const [isValidGroup, group] = validEnumString('group', config.group, groupOptions, 'primary');

  const isValid = isValidLabel && isValidFilterType && isValidSearchMode && isValidGroup;
  const validValue = {
    searchPageConfig: {
      ...label,
      ...filterType,
      ...searchMode,
      ...group,
    },
  };
  return [isValid, validValue];
};

// listingExtendedDataConfig.listingPageConfig
const validListingPageConfig = config => {
  const isUndefined = typeof config === 'undefined';
  if (isUndefined) {
    return [true, {}];
  }
  // Validate: label, isDetail.
  const [isValidLabel, label] = validLabel(config.label);
  const [isValidIsDetail, isDetail] = validBoolean('isDetail', config.isDetail, true);

  const isValid = isValidLabel && isValidIsDetail;
  const validValue = {
    listingPageConfig: {
      ...label,
      ...isDetail,
    },
  };
  return [isValid, validValue];
};

// listingExtendedDataConfig.editListingPageConfig
const validPlaceholderMessage = placeholderMessage => {
  const isUndefined = typeof placeholderMessage === 'undefined';
  const isString = typeof placeholderMessage === 'string';
  const isValid = isUndefined || isString;
  const validValue = isValid ? { placeholderMessage } : {};
  return [isValid, validValue];
};
const validRequiredMessage = requiredMessage => {
  const isUndefined = typeof requiredMessage === 'undefined';
  const isString = typeof requiredMessage === 'string';
  const isValid = isUndefined || isString;
  const validValue = isValid ? { requiredMessage } : {};
  return [isValid, validValue];
};

const validEditListingPageConfig = config => {
  const isUndefined = typeof config === 'undefined';
  if (isUndefined) {
    return [true, {}];
  }
  // Validate: label, placeholderMessage, required, requiredMessage
  const [isValidLabel, label] = validLabel(config.label);
  const [isValidPlaceholder, placeholderMessage] = validPlaceholderMessage(
    config.placeholderMessage
  );
  const [isValidIsRequired, isRequired] = validBoolean('isRequired', config.isRequired, false);
  const [isValidRequiredMessage, requiredMessage] = validRequiredMessage(config.requiredMessage);

  const isValid = isValidLabel && isValidPlaceholder && isValidIsRequired && isValidRequiredMessage;
  const validValue = {
    editListingPageConfig: {
      ...label,
      ...placeholderMessage,
      ...isRequired,
      ...requiredMessage,
    },
  };
  return [isValid, validValue];
};

const validListingExtendedData = (listingExtendedData, processAliasesInUse) => {
  const keys = listingExtendedData.map(d => d.key);
  const scopeOptions = ['public', 'private'];
  const validSchemaTypes = ['enum', 'multi-enum', 'text', 'long', 'boolean'];

  return listingExtendedData.reduce((acc, data) => {
    const schemaType = data.schemaType;

    const validationData = Object.entries(data).reduce(
      (acc, entry) => {
        const [name, value] = entry;

        // Validate each property
        const [isValid, prop] =
          name === 'key'
            ? validKey(value, keys)
            : name === 'scope'
            ? validEnumString('scope', value, scopeOptions, 'public')
            : name === 'includeForProcessAliases'
            ? validProcessAliases(value, processAliasesInUse)
            : name === 'schemaType'
            ? validEnumString('schemaType', value, validSchemaTypes)
            : name === 'schemaOptions'
            ? validSchemaOptions(value, schemaType)
            : name === 'indexForSearch'
            ? validBoolean('indexForSearch', value, false)
            : name === 'searchPageConfig'
            ? validSearchPageConfig(value)
            : name === 'listingPageConfig'
            ? validListingPageConfig(value)
            : name === 'editListingPageConfig'
            ? validEditListingPageConfig(value)
            : [true, value];

        const hasFoundInvalid = !(acc.isValid === false || isValid === false);
        return { config: { ...acc.config, ...prop }, isValid: hasFoundInvalid };
      },
      { config: {}, isValid: true }
    );

    if (validationData.isValid) {
      return [...acc, validationData.config];
    } else {
      console.warn('Unsupported listing extended data configurations detected', data);
      return acc;
    }
  }, []);
};

const getProcessAliasesInUse = transactionTypes => {
  return transactionTypes.map(tt => `${tt.process}/${tt.alias}`);
};
const validListingConfig = (config, transactionTypes) => {
  const listingExtendedData = config?.listingExtendedData || [];
  const processAliasesInUse = getProcessAliasesInUse(transactionTypes);
  return {
    listingExtendedData: validListingExtendedData(listingExtendedData, processAliasesInUse),
  };
};

//////////////////////////////
// Validate Default filters //
//////////////////////////////

const validDatesConfig = config => {
  const label = typeof config.label === 'string' ? config.label : 'Dates';
  const entireRangeAvailable =
    typeof config.entireRangeAvailable === 'boolean' ? config.entireRangeAvailable : true;
  const mode = ['day', 'night'].includes(config.mode) ? config.mode : 'day';
  return { key: 'dates', schemaType: 'dates', label, entireRangeAvailable, mode };
};

const validPriceConfig = config => {
  const label = typeof config.label === 'string' ? config.label : 'Price';
  const min = typeof config.min === 'number' ? config.min : 0;
  const max = typeof config.max === 'number' ? config.max : 1000;
  const step = typeof config.step === 'number' ? config.step : 5;
  return { key: 'price', schemaType: 'price', label, min, max, step };
};

const validDefaultFilters = defaultFilters => {
  return defaultFilters.filter(data => {
    const key = data.key;
    return key === 'dates'
      ? validDatesConfig(data)
      : key === 'price'
      ? validPriceConfig(data)
      : data;
  });
};

//////////////////////////
// Validate sort config //
//////////////////////////

const validSortConfig = config => {
  const active = typeof config.active === 'boolean' ? config.active : true;
  const queryParamName = config.queryParamName || 'sort';
  const relevanceKey = config.relevanceKey || 'relevance';
  const relevanceFilter = config.relevanceFilter || 'keywords';
  const conflictingFilters = config.conflictingFilters || [];
  const optionsRaw = config.options || [];
  const options = optionsRaw.filter(o => !!o.key && !!o.label);
  return { active, queryParamName, relevanceKey, relevanceFilter, conflictingFilters, options };
};

const validSearchConfig = config => {
  const { mainSearchType: mainSearchTypeRaw, defaultFilters, sortConfig, ...rest } = config || {};
  const mainSearchType = ['location', 'keywords'].includes(mainSearchTypeRaw)
    ? mainSearchTypeRaw
    : 'keywords';

  return {
    mainSearchType,
    defaultFilters: validDefaultFilters(defaultFilters),
    sortConfig: validSortConfig(sortConfig),
    ...rest,
  };
};

const validTransactionConfig = config => {
  const { transactionTypes } = config || {};
  // Check what transaction processes this client app supports
  const supportedProcessesInfo = getSupportedProcessesInfo();

  const validTransactionTypes = transactionTypes.reduce((validConfigs, transactionType) => {
    const { type, process: processName, alias, unitType, label, ...rest } = transactionType;
    const isSupportedProcessName = supportedProcessesInfo.find(p => p.name === processName);
    const isSupportedProcessAlias = supportedProcessesInfo.find(p => p.alias === alias);
    const isSupportedUnitType = supportedProcessesInfo.find(p => p.unitTypes.includes(unitType));

    if (isSupportedProcessName && isSupportedProcessAlias && isSupportedUnitType) {
      return [
        ...validConfigs,
        {
          type,
          label,
          process: processName,
          alias,
          unitType,
          // e.g. showStock
          ...rest,
        },
      ];
    }
    console.warn('Unsupported transaction configurations detected', transactionType);
    return validConfigs;
  }, []);
  return { transactionTypes: validTransactionTypes };
};

////////////////////////////////////
// Validate and merge all configs //
////////////////////////////////////

export const mergeConfig = (configAsset = {}, defaultConfigs = {}) => {
  // TODO: defaultConfigs.transaction probably needs to be removed, when config is fetched from assets.
  const validTransactionConfiguration = validTransactionConfig(
    configAsset.transaction || defaultConfigs.transaction
  );

  return {
    ...defaultConfigs,
    branding: mergeBranding(configAsset.branding, defaultConfigs.branding),
    layout: mergeLayouts(configAsset.layout, defaultConfigs.layout),

    // TODO: defaultConfigs.listing probably needs to be removed, when config is fetched from assets.
    listing: validListingConfig(
      configAsset.listing || defaultConfigs.listing,
      validTransactionConfiguration.transactionTypes
    ),
    // TODO: defaultConfigs.search probably needs to be removed, when config is fetched from assets.
    search: validSearchConfig(configAsset.search || defaultConfigs.search),
    transaction: validTransactionConfiguration,
  };
};
