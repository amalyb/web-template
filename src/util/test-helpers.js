import React from 'react';
import mapValues from 'lodash/mapValues';
import Enzyme, { shallow, mount } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import configureStore from '../store';
import { IntlProvider } from '../util/reactIntl';
import defaultConfig from '../config/defaultConfig';
import { ConfigurationProvider } from '../context/configurationContext';
import { RouteConfigurationProvider } from '../context/routeConfigurationContext';
import routeConfiguration from '../routing/routeConfiguration';

// In case you have translated the template and have new translations that
// are missing from the en translations file, the language for the tests can
// be changed here so that there are no missing translation keys in tests.
import messages from '../translations/en.json';

Enzyme.configure({ adapter: new Adapter() });

let undefined;
export const getDefaultConfiguration = () => {
  return {
    ...defaultConfig,
    currency: 'USD',
    facebookAppId: undefined,
    canonicalRootURL: 'http://localhost:3000',
    maps: {
      ...defaultConfig.maps,
      mapboxAccessToken: undefined,
      googleMapsAPIKey: undefined,
    },
    stripe: {
      ...defaultConfig.stripe,
      publishableKey: 'pk_test_',
    },
  };
};

export const getRouteConfiguration = () => {
  const layoutConfig = {
    searchPageVariant: 'map',
    listingPageVariant: 'full-image',
    listingImage: {
      aspectWidth: 400,
      aspectHeight: 400,
      variantPrefix: 'listing-card',
    },
  };
  return routeConfiguration(layoutConfig);
};

// Locale should not affect the tests. We ensure this by providing
// messages with the key as the value of each message.
const testMessages = mapValues(messages, (val, key) => key);

// Provide all the context for components that connect to the Redux
// store, i18n, router, etc.
export const TestProvider = props => {
  const store = configureStore();
  return (
    <ConfigurationProvider value={getDefaultConfiguration()}>
      <RouteConfigurationProvider value={getRouteConfiguration()}>
        <IntlProvider locale="en" messages={testMessages} textComponent="span">
          <BrowserRouter>
            <Provider store={store}>{props.children}</Provider>
          </BrowserRouter>
        </IntlProvider>
      </RouteConfigurationProvider>
    </ConfigurationProvider>
  );
};

// Use Enzyme's shallow rendering to render the given component to a
// JSON structure that can be used in snapshot tests. This doesn't
// render the children within the given component, only a
// representation of the child component and its props.
//
// Useful for snapshot testing components that contain shared
// components. With deep rendering, if the child component changes
// internally, the test for the given component would also fail. This
// avoids the problem by not rendering the full tree but only the
// relevant structure for the given component.
export const renderShallow = component => {
  return toJson(shallow(component));
};

// Fully render the given component to a JSON structure that can be
// used in snapshot tests.
export const renderDeep = component => {
  return toJson(mount(<TestProvider>{component}</TestProvider>), { mode: 'deep' });
};
