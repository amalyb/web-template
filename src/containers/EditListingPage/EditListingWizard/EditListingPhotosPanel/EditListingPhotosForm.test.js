import React from 'react';
import { renderShallow } from '../../../../util/test-helpers';
import { fakeIntl } from '../../../../util/test-data';
import { EditListingPhotosFormComponent } from './EditListingPhotosForm';

const noop = () => null;

describe('EditListingPhotosForm', () => {
  it('matches snapshot', () => {
    const tree = renderShallow(
      <EditListingPhotosFormComponent
        initialValues={{ country: 'US', images: [] }}
        intl={fakeIntl}
        dispatch={noop}
        onImageUpload={v => v}
        onSubmit={v => v}
        saveActionMsg="Save photos"
        stripeConnected={false}
        updated={false}
        ready={false}
        updateInProgress={false}
        disabled={false}
        onRemoveImage={noop}
        listingImageConfig={{ aspectWidth: 1, aspectHeight: 1, variantPrefix: 'listing-card' }}
      />
    );
    expect(tree).toMatchSnapshot();
  });
});
