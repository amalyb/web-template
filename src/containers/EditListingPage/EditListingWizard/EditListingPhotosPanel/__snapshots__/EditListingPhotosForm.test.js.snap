// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditListingPhotosForm matches snapshot 1`] = `
<ReactFinalForm
  disabled={false}
  dispatch={[Function]}
  fetchErrors={null}
  initialValues={
    Object {
      "country": "US",
      "images": Array [],
    }
  }
  intl={
    Object {
      "formatDate": [Function],
      "formatDateTimeRange": [Function],
      "formatMessage": [Function],
      "formatNumber": [Function],
      "formatPlural": [Function],
      "formatRelativeTime": [Function],
      "formatTime": [Function],
      "now": [Function],
    }
  }
  listingImageConfig={
    Object {
      "aspectHeight": 1,
      "aspectWidth": 1,
      "variantPrefix": "listing-card",
    }
  }
  mutators={
    Object {
      "concat": [Function],
      "insert": [Function],
      "move": [Function],
      "pop": [Function],
      "push": [Function],
      "remove": [Function],
      "removeBatch": [Function],
      "shift": [Function],
      "swap": [Function],
      "unshift": [Function],
      "update": [Function],
    }
  }
  onImageUpload={[Function]}
  onRemoveImage={[Function]}
  onSubmit={[Function]}
  ready={false}
  render={[Function]}
  saveActionMsg="Save photos"
  stripeConnected={false}
  updateInProgress={false}
  updated={false}
/>
`;
